package/MDAnalysis/coordinates/PDB.py

```diff
@@ -1078,35 +1077,39 @@ def _write_timestep(self, ts, multiframe=False):
            have been written.)
 
         """
-
-        traj = self.trajectory
         atoms = self.obj.atoms
+        pos = atoms.ts.positions
         if self.convert_units:
-            coor = self.convert_pos_to_native(ts._pos, inplace=False)
-        else:
-            coor = ts._pos
-
-        if hasattr(self.obj, "indices"):
-            coor = coor[self.obj.indices]
-
-        if len(atoms) != len(coor):
-            raise ValueError(
-                "Length of the atoms array is %d, this is different form the Timestep coordinate array %d" % (
-                    len(atoms), len(ts._pos)))
+            pos = self.convert_pos_to_native(pos, inplace=False)
 
         if multiframe:
             self.MODEL(self.frames_written + 1)
 
         for i, atom in enumerate(atoms):
-            # TODO Jan: see description in ATOM for why this check has to be made
-            if not atom.bfactor:
-                atom.bfactor = 0.
-            self.ATOM(serial=i + 1, name=atom.name.strip(), resName=atom.resname.strip(),
-                      resSeq=atom.resid, chainID=atom.segid.strip(), segID=atom.segid.strip(),
-                      tempFactor=atom.bfactor, altLoc=atom.altLoc,
-                      x=coor[i, 0], y=coor[i, 1], z=coor[i, 2])
-            # get bfactor, too, and add to output?
-            # 'element' is auto-guessed from atom.name in ATOM()
+            segid = atom.segid if atom.segid is not "SYSTEM" else " "
+
+            vals = {}
+            vals['serial'] = int(str(i + 1)[-5:])  # check for overflow here?
+            vals['name'] = atom.name[:4]
+            vals['altLoc'] = atom.altLoc[:1] if atom.altLoc is not None else " "
+            vals['resName'] = atom.resname[:4]
+            vals['chainID'] = segid[:1]
+            vals['resSeq'] = int(str(atom.resid)[-4:])
+            vals['iCode'] = " "
+            vals['pos'] = pos[i]  # don't take off atom so conversion works
+            try:
+                occ = atom.occupancy
+            except NoDataError:
+                occ = 1.0
+            vals['occupancy'] = occ if occ is not None else 1.0
+            temp = atom.bfactor
+            vals['tempFactor'] = temp if temp is not None else 0.0
+            vals['segID'] = segid[:4]
+            vals['element'] = guess_atom_element(atom.name.strip())[:2]
+            vals['charge'] = 0
```

Comment:
btw when you want to link to specific lines in a file you should link to the file at a specific version. Otherwise the link will point to different lines of code over time. To create a link bound to a git revision type `y` after marking the line.